### PR TITLE
Charlesmchen/flat button

### PR DESCRIPTION
**Pull request by charlesmchen**
This button is a step towards a "design language":

* Consistent corner rounding.
* Consistent "down state" (usually derived from "up state").
* Consistent font sizes / weights.
* DRY up view code.

PTAL @michaelkirk 

screen shot 2017-09-12 at 10 36 13 am
screen shot 2017-09-12 at 10 30 56 am
screen shot 2017-09-12 at 10 30 52 am
screen shot 2017-09-12 at 10 20 22 am
screen shot 2017-09-12 at 10 17 03 am
screen shot 2017-09-12 at 10 13 06 am
screen shot 2017-09-12 at 10 10 26 am
screen shot 2017-09-12 at 9 40 58 am

### CODE REVIEW

File: Signal/src/util/UIImage+OWS.m [hunk 1]
```diff
@@ -166,4 +168,28 @@ - (UIImage *)resizedImageToFillPixelSize:(CGSize)dstSize
...
+
+    return [self imageWithColor:color size:CGSizeMake(1.f, 1.f)];
+}
+
++ (UIImage *)imageWithColor:(UIColor *)color size:(CGSize)size
+{
+    OWSAssert(color);
+
+    CGRect rect = CGRectMake(0.0f, 0.0f, size.width, size.height);
+    UIGraphicsBeginImageContextWithOptions(rect.size, NO, 1.f);
```
Comment (michaelkirk) [inline Signal/src/util/UIImage+OWS.m]: Can we assert main thread here?

File: Signal/src/views/OWSFlatButton.swift [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    public func setPressedBlock(_ pressedBlock: @escaping () -> Void) {
+        guard self.pressedBlock == nil else {
+            owsFail("Button already has pressed block.")
+            return
+        }
+        self.pressedBlock = pressedBlock
+    }
+
+    internal func buttonPressed() {
+        guard let pressedBlock = pressedBlock else {
```
Comment (michaelkirk) [inline Signal/src/views/OWSFlatButton.swift]: nit: This method could just be `pressedBlock?()`

File: Signal/src/views/OWSFlatButton.swift [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    public class func fontForHeight(_ height: CGFloat) -> UIFont {
+        let fontPointSize = round(height * 0.45)
+        return UIFont.ows_mediumFont(withSize:fontPointSize)!
+    }
+
+    // MARK: Methods
+
+    public func setTitle(title: String, font: UIFont,
+                         titleColor: UIColor ) {
+        guard let button = self.button else {
```
Comment (michaelkirk) [inline Signal/src/views/OWSFlatButton.swift]: nit: Is there a reason `self.button` isn't assigned in init so we can avoid these guards?

File: Signal/src/views/OWSFlatButton.swift [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                                    downColor: UIColor ) {
+        guard let button = self.button else {
+            owsFail("Missing button")
+            return
+        }
+        button.setBackgroundImage(UIImage(color:upColor), for: .normal)
+        button.setBackgroundImage(UIImage(color:downColor), for: .highlighted)
+    }
+
+    public func setBackgroundColors(_ backgroundColor: UIColor ) {
```
Comment (michaelkirk) [inline Signal/src/views/OWSFlatButton.swift]: At the call sites above, I didn't notice the pluralization of this method, and assumed we were just calling UIView's built in `setBackgrondColor:`. 

How about renaming to `setBackgroundColors(upColor:)` /`setBackgroundColorsWithUpColor:` to disambiguate.

File: Signal/src/util/UIImage+OWS.m (hunk 1 above)
Comment (charlesmchen) [reply Signal/src/util/UIImage+OWS.m]: Done.

File: Signal/src/views/OWSFlatButton.swift (hunk 3 above)
Comment (charlesmchen) [reply Signal/src/views/OWSFlatButton.swift]: Done.  I wanted to gather "creation & configuration of views" logic in `createContent()` but pragmatism > purity.

File: Signal/src/views/OWSFlatButton.swift (hunk 4 above)
Comment (charlesmchen) [reply Signal/src/views/OWSFlatButton.swift]: Nice idea.  I also added a `owsFail()` in the `setBackgroundColor:` property setter to catch slip ups.

File: Signal/src/views/OWSFlatButton.swift (hunk 2 above)
Comment (charlesmchen) [reply Signal/src/views/OWSFlatButton.swift]: Done.

**Comment by charlesmchen** [comment]
See: https://github.com/WhisperSystems/Signal-iOS/pull/2549

**Pull request merged**